editor/src/components/canvas/controls/select-mode/select-mode-hooks.ts

```diff
@@ -0,0 +1,407 @@
+  maybeClearHighlightsOnHoverEnd: () => void
+} {
+  const stateRef = useRefEditorState((store) => {
+    return {
+      dispatch: store.dispatch,
+      isResizing: pickIsResizing(store.editor.canvas.dragState),
+      isDragging: pickIsDragging(store.editor.canvas.dragState),
+      selectionEnabled: pickSelectionEnabled(store.editor.canvas, store.editor.keysPressed),
+    }
+  })
+
+  const maybeHighlightOnHover = React.useCallback(
+    (target: TemplatePath): void => {
+      const { dispatch, isDragging, isResizing, selectionEnabled } = stateRef.current
+      if (selectionEnabled && !isDragging && !isResizing) {
+        dispatch([setHighlightedView(target)], 'canvas')
+      }
+    },
+    [stateRef],
+  )
+
+  const maybeClearHighlightsOnHoverEnd = React.useCallback((): void => {
+    const { dispatch, isDragging, isResizing, selectionEnabled } = stateRef.current
+    if (selectionEnabled && !isDragging && !isResizing) {
+      dispatch([clearHighlightedViews()], 'canvas')
+    }
+  }, [stateRef])
+
+  return {
+    maybeHighlightOnHover: maybeHighlightOnHover,
+    maybeClearHighlightsOnHoverEnd: maybeClearHighlightsOnHoverEnd,
+  }
+}
+
+function filterHiddenInstances(
+  hiddenInstances: Array<TemplatePath>,
+  paths: Array<TemplatePath>,
+): Array<TemplatePath> {
+  return paths.filter((path) => hiddenInstances.every((hidden) => !TP.pathsEqual(path, hidden)))
+}
+
+export function getSelectableViews(
+  componentMetadata: JSXMetadata,
+  selectedViews: Array<TemplatePath>,
+  hiddenInstances: Array<TemplatePath>,
+  allElementsDirectlySelectable: boolean,
+): TemplatePath[] {
+  let candidateViews: Array<TemplatePath>
+
+  if (allElementsDirectlySelectable) {
+    candidateViews = MetadataUtils.getAllPaths(componentMetadata)
+  } else {
+    const scenes = MetadataUtils.getAllScenePaths(componentMetadata.components)
+    let rootElementsToFilter: TemplatePath[] = []
+    let dynamicScenesWithFragmentRootViews: ScenePath[] = []
+    utils.fastForEach(scenes, (path) => {
+      const scene = MetadataUtils.findSceneByTemplatePath(componentMetadata.components, path)
+      const rootElements = scene?.rootElements
+      if (
+        MetadataUtils.isSceneTreatedAsGroup(scene) &&
+        rootElements != null &&
+        rootElements.length > 1
+      ) {
+        rootElementsToFilter.push(...rootElements)
+        dynamicScenesWithFragmentRootViews.push(path)
+      }
+    })
+    const allRoots = MetadataUtils.getAllCanvasRootPaths(componentMetadata).filter((rootPath) => {
+      return !rootElementsToFilter.some((path) => TP.pathsEqual(rootPath, path))
+    })
+    let siblings: Array<TemplatePath> = []
+    utils.fastForEach(selectedViews, (view) => {
+      utils.fastForEach(TP.allPaths(view), (ancestor) => {
+        const ancestorChildren = MetadataUtils.getImmediateChildren(componentMetadata, ancestor)
+
+        siblings.push(...ancestorChildren.map((child) => child.templatePath))
```

Comment:
Should probably for each these rather than create an array we immediately bin off.

Reply:
I believe this function was cut and pasted without change from `editor/src/components/canvas/controls/select-mode-control-container.tsx`